File: backend/routers/playlists.py
from fastapi import APIRouter, Depends, HTTPException, status, UploadFile, File
from sqlalchemy.orm import Session
from typing import List, Optional
from .. import models, database, auth
import shutil
from pathlib import Path

router = APIRouter(prefix="/playlists", tags=["playlists"])

# Create playlist
@router.post("/")
async def create_playlist(
    name: str,
    is_public: bool = False,
    db: Session = Depends(database.get_db),
    current_user: models.User = Depends(auth.get_current_user)
):
    playlist = models.Playlist(
        name=name,
        creator_id=current_user.id,
        is_public=is_public
    )
    db.add(playlist)
    db.commit()
    db.refresh(playlist)
    return {
        "id": playlist.id,
        "name": playlist.name,
        "is_public": playlist.is_public,
        "track_count": 0
    }

# List user's playlists
@router.get("/")
async def get_playlists(
    db: Session = Depends(database.get_db),
    current_user: models.User = Depends(auth.get_current_user)
):
    playlists = db.query(models.Playlist).filter(
        models.Playlist.creator_id == current_user.id
    ).all()
    
    return [
        {
            "id": p.id,
            "name": p.name,
            "is_public": p.is_public,
            "track_count": len(p.tracks),
            "thumbnail_path": p.thumbnail_path
        }
        for p in playlists
    ]

# Get playlist details with tracks
@router.get("/{playlist_id}")
async def get_playlist(
    playlist_id: int,
    db: Session = Depends(database.get_db),
    current_user: models.User = Depends(auth.get_current_user_optional)
):
    playlist = db.query(models.Playlist).filter(models.Playlist.id == playlist_id).first()
    if not playlist:
        raise HTTPException(status_code=404, detail="Playlist not found")
    
    # Check access
    if not playlist.is_public and (not current_user or current_user.id != playlist.creator_id):
        raise HTTPException(status_code=403, detail="Access denied")
    
    # Get tracks with order
    playlist_tracks = db.query(models.PlaylistTrack).filter(
        models.PlaylistTrack.playlist_id == playlist_id
    ).order_by(models.PlaylistTrack.order).all()
    
    # Get favorite IDs for current user
    user_favorites = set()
    if current_user:
        favorites = db.query(models.Favorite.track_id).filter(
            models.Favorite.user_id == current_user.id
        ).all()
        user_favorites = {f.track_id for f in favorites}
    
    tracks = []
    for pt in playlist_tracks:
        track = pt.track
        tracks.append({
            "id": track.id,
            "title": track.title,
            "artist": track.artist,
            "album": track.album,
            "duration": track.duration,
            "album_art_path": track.album_art_path,
            "is_public": track.is_public,
            "uploader_name": track.uploader.name if track.uploader else "Unknown",
            "is_favorite": track.id in user_favorites
        })
    
    return {
        "id": playlist.id,
        "name": playlist.name,
        "is_public": playlist.is_public,
        "thumbnail_path": playlist.thumbnail_path,
        "creator_name": playlist.creator.name if playlist.creator else "Unknown",
        "creator_id": playlist.creator_id,
        "track_count": len(tracks),
        "tracks": tracks
    }

# Update playlist
@router.put("/{playlist_id}")
async def update_playlist(
    playlist_id: int,
    name: Optional[str] = None,
    is_public: Optional[bool] = None,
    db: Session = Depends(database.get_db),
    current_user: models.User = Depends(auth.get_current_user)
):
    playlist = db.query(models.Playlist).filter(models.Playlist.id == playlist_id).first()
    if not playlist:
        raise HTTPException(status_code=404, detail="Playlist not found")
    
    if playlist.creator_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not authorized")
    
    if name is not None:
        playlist.name = name
    if is_public is not None:
        playlist.is_public = is_public
    
    db.commit()
    return {"message": "Playlist updated", "id": playlist.id}

# Upload playlist thumbnail
@router.post("/{playlist_id}/thumbnail")
async def upload_playlist_thumbnail(
    playlist_id: int,
    file: UploadFile = File(...),
    db: Session = Depends(database.get_db),
    current_user: models.User = Depends(auth.get_current_user)
):
    playlist = db.query(models.Playlist).filter(models.Playlist.id == playlist_id).first()
    if not playlist:
        raise HTTPException(status_code=404, detail="Playlist not found")
    
    if playlist.creator_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not authorized")
    
    # Save thumbnail
    thumbnail_dir = Path("uploads/playlist_thumbnails")
    thumbnail_dir.mkdir(parents=True, exist_ok=True)
    
    ext = file.filename.split(".")[-1] if "." in file.filename else "jpg"
    thumbnail_filename = f"playlist_{playlist_id}.{ext}"
    thumbnail_path = thumbnail_dir / thumbnail_filename
    
    with open(thumbnail_path, "wb") as f:
        shutil.copyfileobj(file.file, f)
    
    playlist.thumbnail_path = f"uploads/playlist_thumbnails/{thumbnail_filename}"
    db.commit()
    
    return {"message": "Thumbnail uploaded", "thumbnail_path": playlist.thumbnail_path}

# Delete playlist
@router.delete("/{playlist_id}")
async def delete_playlist(
    playlist_id: int,
    db: Session = Depends(database.get_db),
    current_user: models.User = Depends(auth.get_current_user)
):
    playlist = db.query(models.Playlist).filter(models.Playlist.id == playlist_id).first()
    if not playlist:
        raise HTTPException(status_code=404, detail="Playlist not found")
    
    if playlist.creator_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not authorized")
    
    # Delete playlist tracks first
    db.query(models.PlaylistTrack).filter(
        models.PlaylistTrack.playlist_id == playlist_id
    ).delete()
    
    db.delete(playlist)
    db.commit()
    return {"message": "Playlist deleted"}

# Add track to playlist
@router.post("/{playlist_id}/tracks/{track_id}")
async def add_track_to_playlist(
    playlist_id: int,
    track_id: int,
    db: Session = Depends(database.get_db),
    current_user: models.User = Depends(auth.get_current_user)
):
    playlist = db.query(models.Playlist).filter(models.Playlist.id == playlist_id).first()
    if not playlist:
        raise HTTPException(status_code=404, detail="Playlist not found")
    
    if playlist.creator_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not authorized")
    
    track = db.query(models.Track).filter(models.Track.id == track_id).first()
    if not track:
        raise HTTPException(status_code=404, detail="Track not found")
    
    # Check if already in playlist
    existing = db.query(models.PlaylistTrack).filter(
        models.PlaylistTrack.playlist_id == playlist_id,
        models.PlaylistTrack.track_id == track_id
    ).first()
    
    if existing:
        raise HTTPException(status_code=400, detail="Track already in playlist")
    
    # Get max order
    max_order = db.query(models.PlaylistTrack).filter(
        models.PlaylistTrack.playlist_id == playlist_id
    ).count()
    
    playlist_track = models.PlaylistTrack(
        playlist_id=playlist_id,
        track_id=track_id,
        order=max_order
    )
    db.add(playlist_track)
    db.commit()
    
    return {"message": "Track added to playlist"}

# Remove track from playlist
@router.delete("/{playlist_id}/tracks/{track_id}")
async def remove_track_from_playlist(
    playlist_id: int,
    track_id: int,
    db: Session = Depends(database.get_db),
    current_user: models.User = Depends(auth.get_current_user)
):
    playlist = db.query(models.Playlist).filter(models.Playlist.id == playlist_id).first()
    if not playlist:
        raise HTTPException(status_code=404, detail="Playlist not found")
    
    if playlist.creator_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not authorized")
    
    playlist_track = db.query(models.PlaylistTrack).filter(
        models.PlaylistTrack.playlist_id == playlist_id,
        models.PlaylistTrack.track_id == track_id
    ).first()
    
    if not playlist_track:
        raise HTTPException(status_code=404, detail="Track not in playlist")
    
    db.delete(playlist_track)
    db.commit()
    
    return {"message": "Track removed from playlist"}

File: backend-node/routes/playlists.js
const express = require('express');
const router = express.Router();
const multer = require('multer');
const path = require('path');
const fs = require('fs');
const db = require('../db');
const { requireAuth, optionalAuth } = require('../auth');

// Configure multer for thumbnail uploads
const thumbnailStorage = multer.diskStorage({
    destination: (req, file, cb) => {
        const uploadDir = path.join(__dirname, '..', '..', 'uploads', 'playlist_thumbnails');
        if (!fs.existsSync(uploadDir)) {
            fs.mkdirSync(uploadDir, { recursive: true });
        }
        cb(null, uploadDir);
    },
    filename: (req, file, cb) => {
        const ext = path.extname(file.originalname);
        cb(null, `playlist_${req.params.id}${ext}`);
    }
});

const uploadThumbnail = multer({ storage: thumbnailStorage });

// Create playlist
router.post('/', requireAuth, (req, res) => {
    const { name } = req.query;
    const isPublic = req.query.is_public === 'true' ? 1 : 0;

    if (!name) {
        return res.status(400).json({ detail: 'Name is required' });
    }

    const result = db.prepare('INSERT INTO playlists (name, creator_id, is_public) VALUES (?, ?, ?)')
        .run(name, req.user.id, isPublic);

    res.json({
        id: result.lastInsertRowid,
        name,
        is_public: !!isPublic,
        track_count: 0
    });
});

// List user's playlists
router.get('/', requireAuth, (req, res) => {
    const playlists = db.prepare('SELECT * FROM playlists WHERE creator_id = ? ORDER BY id DESC').all(req.user.id);

    const result = playlists.map(p => {
        const trackCount = db.prepare('SELECT COUNT(*) as count FROM playlist_tracks WHERE playlist_id = ?').get(p.id);
        return {
            id: p.id,
            name: p.name,
            is_public: !!p.is_public,
            track_count: trackCount?.count || 0,
            thumbnail_path: p.thumbnail_path
        };
    });

    res.json(result);
});

// Get playlist with tracks
router.get('/:id', optionalAuth, (req, res) => {
    const playlist = db.prepare(`
    SELECT p.*, u.name as creator_name 
    FROM playlists p 
    LEFT JOIN users u ON p.creator_id = u.id 
    WHERE p.id = ?
  `).get(req.params.id);

    if (!playlist) {
        return res.status(404).json({ detail: 'Playlist not found' });
    }

    if (!playlist.is_public && (!req.user || req.user.id !== playlist.creator_id)) {
        return res.status(403).json({ detail: 'Access denied' });
    }

    const tracks = db.prepare(`
    SELECT t.*, u.name as uploader_name, a.album_art_path, ar.name as artist
    FROM playlist_tracks pt
    JOIN tracks t ON pt.track_id = t.id
    LEFT JOIN users u ON t.uploader_id = u.id
    LEFT JOIN albums a ON t.album_id = a.id
    LEFT JOIN artists ar ON t.artist_id = ar.id
    WHERE pt.playlist_id = ?
    ORDER BY pt."order"
  `).all(req.params.id);

    const favoriteIds = req.user
        ? new Set(db.prepare('SELECT track_id FROM favorites WHERE user_id = ?').all(req.user.id).map(f => f.track_id))
        : new Set();

    const tracksWithFavorite = tracks.map(t => ({
        ...t,
        is_public: !!t.is_public,
        is_favorite: favoriteIds.has(t.id)
    }));

    res.json({
        id: playlist.id,
        name: playlist.name,
        is_public: !!playlist.is_public,
        thumbnail_path: playlist.thumbnail_path,
        creator_name: playlist.creator_name,
        creator_id: playlist.creator_id,
        track_count: tracks.length,
        tracks: tracksWithFavorite
    });
});

// Update playlist
router.put('/:id', requireAuth, (req, res) => {
    const playlist = db.prepare('SELECT * FROM playlists WHERE id = ?').get(req.params.id);

    if (!playlist) {
        return res.status(404).json({ detail: 'Playlist not found' });
    }

    if (playlist.creator_id !== req.user.id) {
        return res.status(403).json({ detail: 'Not authorized' });
    }

    const { name, is_public } = req.query;

    if (name) {
        db.prepare('UPDATE playlists SET name = ? WHERE id = ?').run(name, req.params.id);
    }
    if (is_public !== undefined) {
        db.prepare('UPDATE playlists SET is_public = ? WHERE id = ?').run(is_public === 'true' ? 1 : 0, req.params.id);
    }

    res.json({ message: 'Playlist updated', id: parseInt(req.params.id) });
});

// Upload thumbnail
router.post('/:id/thumbnail', requireAuth, uploadThumbnail.single('file'), (req, res) => {
    const playlist = db.prepare('SELECT * FROM playlists WHERE id = ?').get(req.params.id);

    if (!playlist) {
        return res.status(404).json({ detail: 'Playlist not found' });
    }

    if (playlist.creator_id !== req.user.id) {
        return res.status(403).json({ detail: 'Not authorized' });
    }

    if (!req.file) {
        return res.status(400).json({ detail: 'No file uploaded' });
    }

    const thumbnailPath = `uploads/playlist_thumbnails/${req.file.filename}`;
    db.prepare('UPDATE playlists SET thumbnail_path = ? WHERE id = ?').run(thumbnailPath, req.params.id);

    res.json({ message: 'Thumbnail uploaded', thumbnail_path: thumbnailPath });
});

// Delete playlist
router.delete('/:id', requireAuth, (req, res) => {
    const playlist = db.prepare('SELECT * FROM playlists WHERE id = ?').get(req.params.id);

    if (!playlist) {
        return res.status(404).json({ detail: 'Playlist not found' });
    }

    if (playlist.creator_id !== req.user.id) {
        return res.status(403).json({ detail: 'Not authorized' });
    }

    db.prepare('DELETE FROM playlist_tracks WHERE playlist_id = ?').run(req.params.id);
    db.prepare('DELETE FROM playlists WHERE id = ?').run(req.params.id);

    res.json({ message: 'Playlist deleted' });
});

// Add track to playlist
router.post('/:id/tracks/:trackId', requireAuth, (req, res) => {
    const playlist = db.prepare('SELECT * FROM playlists WHERE id = ?').get(req.params.id);

    if (!playlist) {
        return res.status(404).json({ detail: 'Playlist not found' });
    }

    if (playlist.creator_id !== req.user.id) {
        return res.status(403).json({ detail: 'Not authorized' });
    }

    const track = db.prepare('SELECT * FROM tracks WHERE id = ?').get(req.params.trackId);
    if (!track) {
        return res.status(404).json({ detail: 'Track not found' });
    }

    const existing = db.prepare('SELECT * FROM playlist_tracks WHERE playlist_id = ? AND track_id = ?')
        .get(req.params.id, req.params.trackId);

    if (existing) {
        return res.status(400).json({ detail: 'Track already in playlist' });
    }

    const countResult = db.prepare('SELECT COUNT(*) as count FROM playlist_tracks WHERE playlist_id = ?')
        .get(req.params.id);
    const maxOrder = countResult?.count || 0;

    db.prepare('INSERT INTO playlist_tracks (playlist_id, track_id, "order") VALUES (?, ?, ?)')
        .run(req.params.id, req.params.trackId, maxOrder);

    res.json({ message: 'Track added to playlist' });
});

// Remove track from playlist
router.delete('/:id/tracks/:trackId', requireAuth, (req, res) => {
    const playlist = db.prepare('SELECT * FROM playlists WHERE id = ?').get(req.params.id);

    if (!playlist) {
        return res.status(404).json({ detail: 'Playlist not found' });
    }

    if (playlist.creator_id !== req.user.id) {
        return res.status(403).json({ detail: 'Not authorized' });
    }

    const result = db.prepare('DELETE FROM playlist_tracks WHERE playlist_id = ? AND track_id = ?')
        .run(req.params.id, req.params.trackId);

    if (result.changes === 0) {
        return res.status(404).json({ detail: 'Track not in playlist' });
    }

    res.json({ message: 'Track removed from playlist' });
});

module.exports = router;